examples/functional_test.py

```diff
@@ -0,0 +1,279 @@
+    process_classification_outputs,
+)
+from resolver_athena_client.generated.athena.models_pb2 import (
+    ClassificationOutput,
+    ClassifyResponse,
+)
+
+
+async def rate_limited_image_iter(
+    min_interval_ms: int,
+    max_images: int | None = None,
+    counter: list[int] | None = None,
+) -> AsyncIterator[ImageData]:
+    """Generate images with a minimum interval between yields."""
+    last_yield_time = time.time()
+    async for image in iter_images(max_images, counter):
+        elapsed_ms = (time.time() - last_yield_time) * 1000
+        if elapsed_ms < min_interval_ms:
+            await asyncio.sleep((min_interval_ms - elapsed_ms) / 1000)
+        yield image
+        last_yield_time = time.time()
+
+
+def process_errors(
+    logger: logging.Logger, result: ClassifyResponse, current_error_count: int
+) -> int:
+    """Process errors in the result and update the current error count."""
+    if has_output_errors(result):
+        error_summary = get_output_error_summary(result)
+        result_error_count = sum(error_summary.values())
+        current_error_count += result_error_count
+        logger.warning(
+            "Received %d outputs with errors: %s",
+            result_error_count,
+            error_summary,
+        )
+    return current_error_count
+
+
+def dump_classifications(
+    logger: logging.Logger, successful_outputs: list[ClassificationOutput]
+) -> None:
+    """Dump classifications from successful outputs to the logger."""
+    for output in successful_outputs:
+        classifications = {
+            c.label: round(c.weight, 3) for c in output.classifications
+        }
+        logger.debug(
+            "Result [%s]: %s",
+            output.correlation_id[:8],
+            classifications,
+        )
+
+
+async def run_smoke_test(
+    logger: logging.Logger,
+    options: AthenaOptions,
+    credential_helper: CredentialHelper,
+    max_test_images: int,
+    rate_limit_min_interval_ms: int | None = None,
+) -> tuple[int, int, int]:
+    """Run example classification with OAuth credential helper.
+
+    Args:
+        logger: Logger instance for output
+        options: Configuration options for the Athena client
+        credential_helper: OAuth credential helper for authentication
+        max_test_images: Maximum number of test images to generate
+        rate_limit_min_interval_ms: Optional minimum interval in milliseconds
+            between sending images to control rate. If None, sends as fast as
+            possible.
+
+    Returns:
+        Number of requests sent and responses received
+
+    """
+    channel = await create_channel_with_credentials(
+        options.host, credential_helper
+    )
+
+    sent_counter = [0]  # Use list to allow mutation in closure
+    received_count = 0
+    error_count = 0
+
+    img_gen_func = (
+        iter_images
+        if rate_limit_min_interval_ms is None
+        else lambda max_images, sent_ctr: rate_limited_image_iter(
```

Comment:
makes me wonder if we should just make `iter_images` have the rate limit built in. this is basically Functional Programming anyways so maybe it's cool

Reply:
it's only a test function hah if people are sending iter_images in they should just not send in test images